src/vfs2.c

```diff
@@ -1887,48 +1887,43 @@ int vfs2_apply(sqlite3_file *file, struct vfs2_wal_slice stop)
 }
 
 int vfs2_poll(sqlite3_file *file,
-	      dqlite_vfs_frame **frames,
-	      struct vfs2_wal_slice *sl)
+	      dqlite_vfs_frame **frames_out,
+	      struct vfs2_wal_slice *sl_out)
 {
 	struct file *xfile = (struct file *)file;
 	PRE(xfile->flags & SQLITE_OPEN_MAIN_DB);
 	struct entry *e = xfile->entry;
+	PRE(e->shm_locks[WAL_WRITE_LOCK] == 0);
```

Comment:
The write lock has to be off even though no frames were produced? I do think it makes more sense like this but just double checking.

Reply:
Yep, the write lock isn't held because SQLite released it before returning from `sqlite3_step`, regardless of what the transaction looks like. vfs2_poll has the job of re-acquiring it to reflect that there's an uncommitted transaction.